### ISSUE TITLE
How to apply the cuda operator(the kernel functions) designed by ourselves to the model?

### ISSUE THREAD

**Issue by siqiliua** [facebookresearch/CrypTen#388]
I want to apply the kernel functions designed by myself to the crypten model but the onnx can not extrat the cuda operators, what should I do to solve this problem?

**Comment by lvdmaaten** [comment]
If you have custom modules / operators that are not supported by ONNX, I suppose the only thing you can do is build the model graph directly using `crypten.nn`. The `crypten.nn` package works analogous to `torch.nn`. You can implement your own `Module`s and register them in `crypten.nn`; see https://github.com/facebookresearch/CrypTen/blob/main/crypten/nn/module.py#L1033-L1045 for an example of what a `Module` looks like.

Of course, if your `Module.forward` uses custom operators then you are going to have to add those operators (and their gradients) as well. You can do so by registering your own `AutogradFunction`; see https://github.com/facebookresearch/CrypTen/blob/main/crypten/gradients.py#L657-L673 for an example of how the `add` function is registered.
